reader.go

```diff
@@ -31,22 +33,33 @@ func Decode(r io.Reader) (image.Image, error) {
 		read += 4
 		switch string(next) {
 		case "TOC ":
-			read += 4
+			resSize := binary.BigEndian.Uint32(data[read : read+4])
+			read += resSize-4 // size includes header and size fields
 			continue
 		case "icnV":
 			read += 4
 			continue
 		}
+
+		resSize := binary.BigEndian.Uint32(data[read : read+4])
+		read += 4
+		if resSize == 0 {
+			continue // no content, we're not interested
+		}
+
 		if isOsType(string(next)) {
-			iconSize := binary.BigEndian.Uint32(data[read : read+4])
-			read += 4
-			iconData := data[read : read+iconSize]
-			read += iconSize
-			icons = append(icons, iconReader{
-				OsType: osTypeFromID(string(next)),
-				r:      bytes.NewBuffer(iconData),
-			})
+			iconData := data[read : read+resSize-8] // size includes header and size fields
+
+			if bytes.Equal(iconData[:8], jpeg2000header) {
```

Comment:
Perhaps de-indent the `else`, and use:
```go
if bytes.Equal(iconData[:8], jpeg2000header) {
        continue
}
```
This avoids a "useless" empty block.

Reply:
Good point, sorry for the messy code - update inbound :)

Reply:
Sorry it was more of a change than implied - the `+= resSize` needed to move ahead of the loop to avoid duplication.
Happy that it looks much cleaner though, thanks.